🐁 💪 Add MuRP

cthoyt opened the pull request:
This PR is a (first/bad) attempt to incorporate MuRP (for #343). It heavily borrows from Balažević's original implementation at https://github.com/ibalazevic/multirelational-poincare and does its best to give attribution. This PR does the following:

- Implements the Riemannian Stochastic Gradient Descent (SGD) optimizer
- Implements MuRP functional form (differs from original implementation by refactoring redundant code into helper functions - this actually makes it look very similar to MuRE!)
- Implements MuRP interaction module
- Implements MuRP model

Still to do:
- [ ] Refactor parameter tagging used in Riemannian SGD to be more generalizable
- [ ] Consider adding information about default / allowed optimizers for each model?

cthoyt committed:
Add MuRP

References #343

still need to deal with flagging which tensors should get treated as poincare

cthoyt committed:
Update __init__.py

cthoyt committed:
Update __init__.py

mberr commented on src/pykeen/optimizers/rsgd.py:
```diff
@@ -0,0 +1,57 @@
+]
+
+
+class RiemannianSGD(Optimizer):
```
can this be subclassed from [`SGD`](https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html)?

cthoyt replied:
If you look into the implementation of SGD, it's pretty monolithic. I don't see where we could effectively make use of parts that are already implemented. Plus, there's tons of magic in it that I don't understand

mberr commented on src/pykeen/nn/functional.py:
```diff
@@ -956,6 +956,34 @@ def mure_interaction(
+
+def _norm(h):
+    h_norm = torch.norm(h, 2, dim=-1, keepdim=True)
+    return torch.where(h_norm >= 1, h / (h_norm - 1e-5), h)
```
Is this equivalent to `pykeen.utils.clamp_norm`?

cthoyt replied:
Hmm... I will write some code to test if they're the same since I dont want to make a mistake in reasoning about it

cthoyt replied:
Survey says: **yes**

```python
import torch

from pykeen.utils import clamp_norm


def _norm(h, eps=1e-5):
    h_norm = torch.norm(h, 2, dim=-1, keepdim=True)
    return torch.where(h_norm >= 1, h / (h_norm - eps), h)


def main():
    for _ in range(200):
        for eps in [1e-5, 1e-8, 1e-10]:
            for initializer in [torch.nn.init.normal_, torch.nn.init.uniform_]:
                x = torch.ones(50)
                x.data = initializer(x)

                pykeen_clamp = clamp_norm(x, maxnorm=1, eps=eps)
                murp_clamp = _norm(x, eps=eps)
                assert torch.allclose(pykeen_clamp, murp_clamp)


if __name__ == '__main__':
    main()
```

cthoyt committed:
Update to use pykeen implementation of clamp_norm

cthoyt committed:
Pass flake8

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -990,6 +990,36 @@ def get_expected_norm(
         raise TypeError(f"norm not implemented for {type(p)}: {p}")
 
 
+def artanh(x):
```
is this the same as [`torch.atanh`](https://pytorch.org/docs/stable/generated/torch.atanh.html#torch.atanh)?

cthoyt replied:
Should be: https://mathworld.wolfram.com/InverseHyperbolicTangent.html

mberr replied:
yes, cf. [here](https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions#Inverse_hyperbolic_tangent)

mberr commented on src/pykeen/utils.py:
```diff
@@ -990,6 +990,36 @@ def get_expected_norm(
+
+
+def p_sum(x, y):
+    sqxnorm = torch.clamp(torch.sum(x * x, dim=-1, keepdim=True), 0, 1 - 1e-5)
```
Shorter:
```python
sqxnorm = x.pow(2).sum(dim=-1, keepdim=True).clamp_max(1 - 1e-5)
```

cthoyt replied:
Updated in 33d8dbcc

mberr commented on src/pykeen/nn/functional.py:
```diff
@@ -956,6 +956,49 @@ def mure_interaction(
+    """
+    h = clamp_norm(h, maxnorm=1)
+    t = clamp_norm(t, maxnorm=1)
+    r_mat = clamp_norm(r_mat, maxnorm=1)
```
isn't `r_mat` a matrix? By default, the norm is computed along the last dimension only.

cthoyt replied:
yes, r_mat is a matrix. here's the code from the original implementation, which seems to be doing the same as the non-matrix ones:

https://github.com/ibalazevic/multirelational-poincare/blob/34523a61ca7867591fd645bfb0c0807246c08660/model.py#L30-L31

mberr replied:
`rvh` is the vector in the original code, i.e., `r_vec` for us; `Ru` is the matrix, i.e., `r_mat` for us.

cthoyt replied:
oops i have made a mistake then

cthoyt replied:
Fixed in b15914c0

mberr replied:
Now that I took a closer look at the paper and the code, both are vectors, since the matrix is modelled as a diagonal matrix.

cthoyt committed:
Use torch implementation instead

cthoyt committed:
Refactor redundant code in calculation of clamped square norms

cthoyt committed:
Fix functional form

mberr committed:
make MuRP's relation matrix diagonal

mberr committed:
adjust MuRPInteraction's relation_shape

mberr committed:
Merge branch 'master' into add-murp

mberr committed:
simplify code for p_exp_map and p_log_map

mberr committed:
Merge remote-tracking branch 'origin/add-murp' into add-murp

mberr committed:
simplify code for other p_* methods

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -990,14 +990,14 @@ def get_expected_norm(
-    return torch.tanh(normv) * v / normv
+def p_exp_map(v: torch.FloatTensor, eps: float = 1.0e-10) -> torch.FloatTensor:
+    v_norm = v.norm(p=2, dim=-1, keepdim=True)
+    return v_norm.tanh() * v / v_norm.clamp_min(eps)
```
is this because you don't need to clamp before running tanh()?

mberr replied:
As far as I saw it, the clamping was necessary to avoid division by zero. However, `v_norm` was also used as nominator, where there is no risk of numerical problems. So now, the clamping is only applied in the denominator.

cthoyt committed:
Merge branch 'master' into add-murp

cthoyt committed:
Add test

cthoyt committed:
Hack optimizer, training loop, and pipeline to pass information about which embeddings are poincare

I'm not at all happy with this, but there's very little information actually passed to the optimizer, so it's hard to recover it. That's why i build an object ID to tensor mapping, then pre-labeled it inside the model itself.

mberr commented on src/pykeen/optimizers/rsgd.py:
```diff
@@ -20,10 +22,17 @@
     """A variant of :class:`torch.optim.SGD` generalized for riemannian manifolds."""
 
-    def __init__(self, params, lr=required, param_names=None):
+    def __init__(self, params: Sequence[Parameter], param_names: Sequence[str], poincare: Set[str], lr=required):
```
iirc, `params` and `param_names` should match in length?

Then why not provide `params: Mapping[str, Parameter]` instead?

mberr commented on src/pykeen/optimizers/rsgd.py:
```diff
@@ -20,10 +22,17 @@
         defaults = dict(lr=lr)
         super().__init__(params, defaults)
-        self.param_names = [] if param_names is None else param_names
+        self._pykeen_extras = {
```
Why are these parameters saved in a dictionary rather than two attributes?

mberr commented on src/pykeen/optimizers/rsgd.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def euclidean_update(p, d_p, lr):
+    p.data = p.data - lr * d_p
```
The result is also assigned to `p.data`. So this is redundant?

mberr commented on src/pykeen/optimizers/rsgd.py:
```diff
@@ -0,0 +1,69 @@
+
+def poincare_update(p, d_p, lr):
+    v = -lr * d_p
+    p.data = full_p_exp_map(p.data, v)
```
same here

mberr commented on src/pykeen/optimizers/rsgd.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def poincare_grad(p, d_p):
+    p_sqnorm = torch.clamp(torch.sum(p.data ** 2, dim=-1, keepdim=True), 0, 1 - 1e-5)
```
```suggestion
    p_sqnorm = p.data.pow(2).sum(dim=-1, keepdim=True).clamp(min=0, max=1 - 1.0e-05)
```

mberr replied:
The `1 - 1.0e-05` is part of the [Poincare disk model](https://en.wikipedia.org/wiki/Poincar%C3%A9_disk_model), right? I am not very deep into the topic, but is RiemannianSGD restricted to this model? At least when skimming [this text](https://andbloch.github.io/Stochastic-Gradient-Descent-on-Riemannian-Manifolds/) it seemed like the method is more general.

So maybe we should document this limitation(?) somewhere.

btw, the last mentioned blog post refers to a library https://github.com/geoopt/geoopt which seems to implement RiemannianSGD along other optimizers in PyTorch.

cthoyt committed:
Update src/pykeen/optimizers/rsgd.py

Co-authored-by: Max Berrendorf <berrendorf@dbs.ifi.lmu.de>
